### PR TITLE
Added Cloudwatch Alarm for DSS Batch Job

### PR DESCRIPTION
**Pull request by ChrisMKinsella**
@marcus-bcl the plan below was created using a hardcoded queue arn as until we apply the batch/dss the output isn't available to use.

```An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create
  ~ update in-place

Terraform will perform the following actions:

  ~ aws_cloudwatch_dashboard.delius_service_health
      dashboard_body:                                                   "{\"start\":\"-PT6H\",\"widgets\":[{\"type\":\"log\",\"x\":9,\"y\":0,\"width\":15,\"height\":6,\"properties\":{\"query\":\"SOURCE 'tf-eu-west-2-hmpps-delius-prod/weblogic-ndelius' | parse '* - * [*] \\\"* * *\\\" * *' as host, identity, dateTimeString, httpVerb, URL, protocol, statusCode, bytes\\n| filter httpVerb='GET' and URL like /(?i)\\\\.jsp$/\\n| stats count(*) as Requests by URL\\n| sort Requests desc\\n| limit 10\",\"region\":\"eu-west-2\",\"stacked\":false,\"title\":\"Top Pages\",\"view\":\"table\"}},{\"type\":\"log\",\"x\":18,\"y\":6,\"width\":6,\"height\":6,\"properties\":{\"query\":\"SOURCE 'tf-eu-west-2-hmpps-delius-prod/weblogic-ndelius' | SOURCE 'tf-eu-west-2-hmpps-delius-prod/weblogic-interface' | SOURCE 'tf-eu-west-2-hmpps-delius-prod/weblogic-spg' | parse 'FATAL * * * - * Error Code: *' as date, timestamp, exception, user, ErrorCode\\n| filter @logStream like /NDeliusLog_log/ and ErrorCode like /.*/ and @message not like /at .*\\\\(.*:.*\\\\)/ and @message not like /Caused by: /\\n| stats count(*) as Count by ErrorCode\\n| sort by Count desc\\n| limit 10\",\"region\":\"eu-west-2\",\"stacked\":false,\"title\":\"Top Errors\",\"view\":\"table\"}},{\"type\":\"metric\",\"x\":3,\"y\":0,\"width\":6,\"height\":3,\"properties\":{\"metrics\":[[\"AWS/ApplicationELB\",\"RequestCount\",\"LoadBalancer\",\"app/del-prod-ndelius-alb/0206105caae02a41\",{\"id\":\"m1\",\"stat\":\"Sum\",\"label\":\"NDelius\",\"yAxis\":\"left\"}]],\"view\":\"timeSeries\",\"stacked\":true,\"region\":\"eu-west-2\",\"title\":\"Requests\",\"period\":60,\"liveData\":false}},{\"type\":\"metric\",\"x\":3,\"y\":3,\"width\":6,\"height\":3,\"properties\":{\"metrics\":[[\"AWS/ApplicationELB\",\"TargetResponseTime\",\"LoadBalancer\",\"app/del-prod-ndelius-alb/0206105caae02a41\",{\"id\":\"m2\",\"label\":\"NDelius\",\"stat\":\"Average\"}],[{\"expression\":\"ANOMALY_DETECTION_BAND(m2, 2)\",\"label\":\"NDelius (expected)\",\"id\":\"ad1\",\"color\":\"#666666\",\"region\":\"eu-west-2\"}]],\"view\":\"timeSeries\",\"stacked\":false,\"region\":\"eu-west-2\",\"title\":\"Response Time (seconds)\",\"period\":60,\"liveData\":false}},{\"type\":\"metric\",\"x\":0,\"y\":12,\"width\":12,\"height\":3,\"properties\":{\"metrics\":[[\"AWS/EC2\",\"CPUUtilization\",\"InstanceId\",\"i-0e04db1660e98a78e\",{\"label\":\"Database\"}],[\"...\",\"AutoScalingGroupName\",\"delius-prod-ndelius\",{\"label\":\"WebLogic\"}],[\"...\",\"delius-prod-ldap\",{\"label\":\"LDAP\",\"stat\":\"Maximum\"}]],\"view\":\"timeSeries\",\"stacked\":false,\"region\":\"eu-west-2\",\"period\":300,\"title\":\"CPU Load\",\"legend\":{\"position\":\"right\"},\"liveData\":true}},{\"type\":\"log\",\"x\":9,\"y\":6,\"width\":9,\"height\":6,\"properties\":{\"query\":\"SOURCE 'tf-eu-west-2-hmpps-delius-prod/weblogic-ndelius' | SOURCE 'tf-eu-west-2-hmpps-delius-prod/weblogic-interface' | SOURCE 'tf-eu-west-2-hmpps-delius-prod/weblogic-spg' | parse 'FATAL * * * - * Error Code: *' as date, timestamp, exception, user, ErrorCode\\n| filter @logStream like /NDeliusLog_log/ and ErrorCode like /.*/ and @message not like /at .*\\\\(.*:.*\\\\)/ and @message not like /Caused by: /\\n| stats count(*)/5 as Count by bin(5m)\",\"region\":\"eu-west-2\",\"stacked\":false,\"title\":\"Errors\",\"view\":\"timeSeries\"}},{\"type\":\"metric\",\"x\":0,\"y\":0,\"width\":3,\"height\":6,\"properties\":{\"metrics\":[[\"WebLogic\",\"OpenSessions\",\"AutoScalingGroupName\",\"delius-prod-ndelius\",{\"label\":\"Current\",\"stat\":\"Sum\"}],[\".\",\"TotalSessions\",\".\",\".\",{\"label\":\"Total (today)\",\"stat\":\"Sum\"}]],\"view\":\"singleValue\",\"region\":\"eu-west-2\",\"title\":\"Sessions\",\"period\":60}},{\"type\":\"metric\",\"x\":0,\"y\":6,\"width\":3,\"height\":6,\"properties\":{\"metrics\":[[\"WebLogic\",\"InboundQueueSize\",\"AutoScalingGroupName\",\"delius-prod-spg\",{\"stat\":\"Sum\",\"label\":\"Inbound\"}],[\".\",\"OutboundQueueSize\",\".\",\".\",{\"stat\":\"Sum\",\"label\":\"Outbound\"}]],\"view\":\"singleValue\",\"region\":\"eu-west-2\",\"title\":\"SPG Queue Size\",\"period\":60}},{\"type\":\"metric\",\"x\":3,\"y\":6,\"width\":6,\"height\":6,\"properties\":{\"metrics\":[[{\"expression\":\"RATE(METRICS())\",\"label\":\"\",\"region\":\"eu-west-2\"}],[\"WebLogic\",\"OutboundQueueDispatchCount\",\"AutoScalingGroupName\",\"delius-prod-spg\",{\"label\":\"Outbound\",\"visible\":false}],[\".\",\"InboundQueueDispatchCount\",\".\",\".\",{\"label\":\"Inbound\",\"visible\":false}]],\"view\":\"timeSeries\",\"stacked\":true,\"region\":\"eu-west-2\",\"period\":60,\"title\":\"SPG Messages\",\"legend\":{\"position\":\"bottom\"},\"liveData\":false,\"yAxis\":{\"left\":{\"showUnits\":false,\"min\":0}},\"stat\":\"Sum\"}},{\"type\":\"metric\",\"x\":12,\"y\":12,\"width\":12,\"height\":3,\"properties\":{\"metrics\":[[{\"expression\":\"m1-m2\",\"label\":\"Used\",\"id\":\"e1\",\"region\":\"eu-west-2\",\"yAxis\":\"left\"}],[\"WebLogic\",\"HeapSizeCurrent\",\"AutoScalingGroupName\",\"delius-prod-ndelius\",{\"id\":\"m1\",\"label\":\"Max\",\"color\":\"#d62728\"}],[\".\",\"HeapFreeCurrent\",\".\",\".\",{\"id\":\"m2\",\"visible\":false,\"label\":\"Free\"}]],\"view\":\"timeSeries\",\"stacked\":false,\"region\":\"eu-west-2\",\"title\":\"WebLogic Memory Usage\",\"period\":60,\"legend\":{\"position\":\"right\"},\"yAxis\":{\"left\":{\"min\":0}},\"stat\":\"Sum\"}}]}" => "{\"start\":\"-PT6H\",\"widgets\":[{\"height\":6,\"properties\":{\"query\":\"SOURCE 'tf-eu-west-2-hmpps-delius-prod/weblogic-ndelius' | parse '* - * [*] \\\"* * *\\\" * *' as host, identity, dateTimeString, httpVerb, URL, protocol, statusCode, bytes\\n| filter httpVerb='GET' and URL like /(?i)\\\\.jsp$/\\n| stats count(*) as Requests by URL\\n| sort Requests desc\\n| limit 10\",\"region\":\"eu-west-2\",\"stacked\":false,\"title\":\"Top Pages\",\"view\":\"table\"},\"type\":\"log\",\"width\":15,\"x\":9,\"y\":0},{\"height\":6,\"properties\":{\"query\":\"SOURCE 'tf-eu-west-2-hmpps-delius-prod/weblogic-ndelius' | SOURCE 'tf-eu-west-2-hmpps-delius-prod/weblogic-interface' | SOURCE 'tf-eu-west-2-hmpps-delius-prod/weblogic-spg' | parse 'FATAL * * * - * Error Code: *' as date, timestamp, exception, user, ErrorCode\\n| filter @logStream like /NDeliusLog_log/ and ErrorCode like /.*/ and @message not like /at .*\\\\(.*:.*\\\\)/ and @message not like /Caused by: /\\n| stats count(*) as Count by ErrorCode\\n| sort by Count desc\\n| limit 10\",\"region\":\"eu-west-2\",\"stacked\":false,\"title\":\"Top Errors\",\"view\":\"table\"},\"type\":\"log\",\"width\":6,\"x\":18,\"y\":6},{\"height\":3,\"properties\":{\"liveData\":false,\"metrics\":[[\"AWS/ApplicationELB\",\"RequestCount\",\"LoadBalancer\",\"app/del-prod-ndelius-alb/0206105caae02a41\",{\"id\":\"m1\",\"label\":\"NDelius\",\"stat\":\"Sum\",\"yAxis\":\"left\"}]],\"period\":60,\"region\":\"eu-west-2\",\"stacked\":true,\"title\":\"Requests\",\"view\":\"timeSeries\"},\"type\":\"metric\",\"width\":6,\"x\":3,\"y\":0},{\"height\":3,\"properties\":{\"liveData\":false,\"metrics\":[[\"AWS/ApplicationELB\",\"TargetResponseTime\",\"LoadBalancer\",\"app/del-prod-ndelius-alb/0206105caae02a41\",{\"id\":\"m2\",\"label\":\"NDelius\",\"stat\":\"Average\"}],[{\"color\":\"#666666\",\"expression\":\"ANOMALY_DETECTION_BAND(m2, 2)\",\"id\":\"ad1\",\"label\":\"NDelius (expected)\",\"region\":\"eu-west-2\"}]],\"period\":60,\"region\":\"eu-west-2\",\"stacked\":false,\"title\":\"Response Time (seconds)\",\"view\":\"timeSeries\"},\"type\":\"metric\",\"width\":6,\"x\":3,\"y\":3},{\"height\":3,\"properties\":{\"legend\":{\"position\":\"right\"},\"liveData\":true,\"metrics\":[[\"AWS/EC2\",\"CPUUtilization\",\"InstanceId\",\"i-0e04db1660e98a78e\",{\"label\":\"Database\"}],[\"...\",\"AutoScalingGroupName\",\"delius-prod-ndelius\",{\"label\":\"WebLogic\"}],[\"...\",\"delius-prod-ldap\",{\"label\":\"LDAP\",\"stat\":\"Maximum\"}]],\"period\":300,\"region\":\"eu-west-2\",\"stacked\":false,\"title\":\"CPU Load\",\"view\":\"timeSeries\"},\"type\":\"metric\",\"width\":12,\"x\":0,\"y\":12},{\"height\":6,\"properties\":{\"query\":\"SOURCE 'tf-eu-west-2-hmpps-delius-prod/weblogic-ndelius' | SOURCE 'tf-eu-west-2-hmpps-delius-prod/weblogic-interface' | SOURCE 'tf-eu-west-2-hmpps-delius-prod/weblogic-spg' | parse 'FATAL * * * - * Error Code: *' as date, timestamp, exception, user, ErrorCode\\n| filter @logStream like /NDeliusLog_log/ and ErrorCode like /.*/ and @message not like /at .*\\\\(.*:.*\\\\)/ and @message not like /Caused by: /\\n| stats count(*)/5 as Count by bin(5m)\",\"region\":\"eu-west-2\",\"stacked\":false,\"title\":\"Errors\",\"view\":\"timeSeries\"},\"type\":\"log\",\"width\":9,\"x\":9,\"y\":6},{\"height\":6,\"properties\":{\"metrics\":[[\"WebLogic\",\"OpenSessions\",\"AutoScalingGroupName\",\"delius-prod-ndelius\",{\"label\":\"Current\",\"stat\":\"Sum\"}],[\".\",\"TotalSessions\",\".\",\".\",{\"label\":\"Total (today)\",\"stat\":\"Sum\"}]],\"period\":60,\"region\":\"eu-west-2\",\"title\":\"Sessions\",\"view\":\"singleValue\"},\"type\":\"metric\",\"width\":3,\"x\":0,\"y\":0},{\"height\":6,\"properties\":{\"metrics\":[[\"WebLogic\",\"InboundQueueSize\",\"AutoScalingGroupName\",\"delius-prod-spg\",{\"label\":\"Inbound\",\"stat\":\"Sum\"}],[\".\",\"OutboundQueueSize\",\".\",\".\",{\"label\":\"Outbound\",\"stat\":\"Sum\"}]],\"period\":60,\"region\":\"eu-west-2\",\"title\":\"SPG Queue Size\",\"view\":\"singleValue\"},\"type\":\"metric\",\"width\":3,\"x\":0,\"y\":6},{\"height\":6,\"properties\":{\"legend\":{\"position\":\"bottom\"},\"liveData\":false,\"metrics\":[[{\"expression\":\"RATE(METRICS())\",\"label\":\"\"}],[\"WebLogic\",\"InboundQueueDispatchCount\",\"AutoScalingGroupName\",\"delius-prod-spg\",{\"label\":\"Inbound\",\"visible\":false}],[\".\",\"OutboundQueueDispatchCount\",\".\",\".\",{\"label\":\"Outbound\",\"visible\":false}]],\"period\":60,\"region\":\"eu-west-2\",\"stacked\":true,\"stat\":\"Sum\",\"title\":\"SPG Messages\",\"view\":\"timeSeries\",\"yAxis\":{\"left\":{\"min\":0,\"showUnits\":false}}},\"type\":\"metric\",\"width\":6,\"x\":3,\"y\":6},{\"height\":3,\"properties\":{\"legend\":{\"position\":\"right\"},\"metrics\":[[{\"expression\":\"m1-m2\",\"id\":\"e1\",\"label\":\"Used\",\"region\":\"eu-west-2\",\"yAxis\":\"left\"}],[\"WebLogic\",\"HeapSizeCurrent\",\"AutoScalingGroupName\",\"delius-prod-ndelius\",{\"color\":\"#d62728\",\"id\":\"m1\",\"label\":\"Max\"}],[\".\",\"HeapFreeCurrent\",\".\",\".\",{\"id\":\"m2\",\"label\":\"Free\",\"visible\":false}]],\"period\":60,\"region\":\"eu-west-2\",\"stacked\":false,\"stat\":\"Sum\",\"title\":\"WebLogic Memory Usage\",\"view\":\"timeSeries\",\"yAxis\":{\"left\":{\"min\":0}}},\"type\":\"metric\",\"width\":12,\"x\":12,\"y\":12}]}"

  + aws_cloudwatch_event_rule.dss_failure_event_rule
      id:                                                               <computed>
      arn:                                                              <computed>
      description:                                                      "delius-prod-dss-batch-job-failure"
      event_pattern:                                                    "{\"detail\":{\"jobQueue\":\"arn:aws:batch:eu-west-2:050243167760:job-queue/delius-prod-ndelius-queue\",\"status\":[\"FAILED\"]},\"detail-type\":[\"Batch Job State Change\"],\"source\":[\"aws.batch\"]}"
      is_enabled:                                                       "true"
      name:                                                             "delius-prod-dss-batch-job-failure"

  + aws_cloudwatch_event_target.sns
      id:                                                               <computed>
      arn:                                                              "arn:aws:sns:eu-west-2:050243167760:delius-prod-delius-core-alarm-notification"
      rule:                                                             "delius-prod-dss-batch-job-failure"
      target_id:                                                        "SendToSNS"

  + aws_cloudwatch_metric_alarm.inbound_queue_rate_critical_alarm
      id:                                                               <computed>
      actions_enabled:                                                  "true"
      alarm_actions.#:                                                  "1"
      alarm_actions.953424210:                                          "arn:aws:sns:eu-west-2:050243167760:delius-prod-delius-core-alarm-notification"
      alarm_description:                                                "Inbound queue size increased at an elevated rate for over 10 minutes. Failover to standby may be required."
      alarm_name:                                                       "delius-prod-activemq-inbound-queue-rate-cwa--critical"
      arn:                                                              <computed>
      comparison_operator:                                              "GreaterThanThreshold"
      evaluate_low_sample_count_percentiles:                            <computed>
      evaluation_periods:                                               "10"
      metric_query.#:                                                   "2"
      metric_query.1972577184.expression:                               "RATE(m1)"
      metric_query.1972577184.id:                                       "e1"
      metric_query.1972577184.label:                                    "InboundQueueRate"
      metric_query.1972577184.metric.#:                                 "0"
      metric_query.1972577184.return_data:                              "true"
      metric_query.3866455510.expression:                               ""
      metric_query.3866455510.id:                                       "m1"
      metric_query.3866455510.label:                                    ""
      metric_query.3866455510.metric.#:                                 "1"
      metric_query.3866455510.metric.0.dimensions.%:                    "1"
      metric_query.3866455510.metric.0.dimensions.AutoScalingGroupName: "delius-prod-spg"
      metric_query.3866455510.metric.0.metric_name:                     "InboundQueueSize"
      metric_query.3866455510.metric.0.namespace:                       "WebLogic"
      metric_query.3866455510.metric.0.period:                          "60"
      metric_query.3866455510.metric.0.stat:                            "Sum"
      metric_query.3866455510.metric.0.unit:                            ""
      metric_query.3866455510.return_data:                              "false"
      ok_actions.#:                                                     "1"
      ok_actions.953424210:                                             "arn:aws:sns:eu-west-2:050243167760:delius-prod-delius-core-alarm-notification"
      threshold:                                                        "1"
      treat_missing_data:                                               "missing"

  ~ aws_cloudwatch_metric_alarm.inbound_queue_size_warning_alarm
      alarm_description:                                                "Inbound queue size exceeded 25 for 30 minutes." => "Inbound queue size exceeded 30 for 30 minutes."
      threshold:                                                        "25" => "30"

  ~ aws_lambda_function.notify_slack
      last_modified:                                                    "2020-09-02T18:46:27.336+0000" => <computed>
      source_code_hash:                                                 "SXv858ECRntCBre511e+QyGq3/rsEiQ+7u4NUEiOf5A=" => "muK4VH4atTSqLcL+K+wikYnYhCAxEZsmiJJB55yV03k="


Plan: 3 to add, 3 to change, 0 to destroy.

------------------------------------------------------------------------

This plan was saved to: delius-prod.plan
```

### CODE REVIEW

File: monitoring/templates/lambda/notify-slack-batch.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    }
+    
+    console.log("severity:" + severity);
+    console.log("icon_emoji:" + icon_emoji);
+    console.log("sendSlackNotification:" + sendSlackNotification);
+    
+    // let jobId = eventMessage.detail.jobId;
+    // let jobName = eventMessage.detail.jobName;
+    // let jobQueue = eventMessage.detail.jobQueue;
+    let logsPath = "https://eu-west-2.console.aws.amazon.com/cloudwatch/home?region=eu-west-2#logsV2:log-groups/log-group/$252Faws$252Flambda$252F" + environment_name + "-notify-delius-core-slack-channel-batch";
```
Comment (marcus-bcl) [inline monitoring/templates/lambda/notify-slack-batch.js:45]: Can we get some of these vars from terraform, as this .js file is templated in [data.tf](https://github.com/ministryofjustice/hmpps-delius-core-terraform/pull/375/files#diff-1a4d37934355f9821b17b6af88e2d578R97-R105)?
```suggestion
    let logsPath = "https://eu-west-2.console.aws.amazon.com/cloudwatch/home?region=eu-west-2#logsV2:log-groups/log-group/$252Faws$252Flambda$252F${lambda_name}";
```

File: monitoring/templates/lambda/notify-slack-batch.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    // console.log("statusReason:" + statusReason);
+    
+    let envvars = eventMessage.detail.container.environment;
+    for (var current in envvars) {
+      // console.log(envvars[current].name);
+      if(envvars[current].name === "DSS_ENVIRONMENT") {
+          environment_name = envvars[current].value;
+          console.log("setting environment_name to " + environment_name);
+      }
+    }
```
Comment (marcus-bcl) [inline monitoring/templates/lambda/notify-slack-batch.js:60]: Can we get some of these vars from terraform, as this .js file is templated in [data.tf](https://github.com/ministryofjustice/hmpps-delius-core-terraform/pull/375/files#diff-1a4d37934355f9821b17b6af88e2d578R97-R105)?
```suggestion
    environment_name="${environment_name}";
```

**Review by marcus-bcl — APPROVED** [review]
Approved with comments

File: monitoring/lambda.tf [hunk 3]
```diff
@@ -13,4 +13,29 @@ resource "aws_lambda_permission" "with_sns" {
...
+
+resource "aws_lambda_function" "notify_slack_batch" {
+  runtime          = "nodejs12.x"
+  role             = "${data.aws_iam_role.lambda_exec_role.arn}"
+  filename         = "${data.archive_file.lambda_handler_zip_batch.output_path}"
+  function_name    = "${local.lambda_name_batch}"
+  handler          = "notify-slack-batch.handler"
+  source_code_hash = "${base64sha256(file("${data.archive_file.lambda_handler_zip_batch.output_path}"))}"
+
+  environment {
```
Comment (ChrisMKinsella) [inline monitoring/lambda.tf:27]: I've added the environment variable into Lambda as we know it at deploy time.

File: monitoring/templates/lambda/notify-slack-batch.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+exports.handler = function(event, context) {
+    console.log(JSON.stringify(event, null, 2));
+    const londonTime = new Date().toLocaleString("en-GB", {timeZone: "Europe/London"});
+    const now = new Date(londonTime).getHours();
+    if (now >= +"${quiet_period_start_hour}" && now < +"${quiet_period_end_hour}") {
+        console.log("In quiet period, dismissing alarm");
+        return;
+    }
+
+    let environment_name = process.env.ENVIRONMENT_NAME;
```
Comment (ChrisMKinsella) [inline monitoring/templates/lambda/notify-slack-batch.js:13]: added this line to get environment_name from Env Vars

**Pull request merged**